README.md
<!--
  Title: CONSUL
  Description: Citizen Participation and Open Government Application
  Keywords: democracy, citizen participation, eparticipation, debates, proposals, voting, consultations, crowdlaw, participatory budgeting
-->

![CONSUL logo](https://raw.githubusercontent.com/consul/consul/master/public/consul_logo.png)

# CONSUL

Citizen Participation and Open Government Application

[![Build Status](https://travis-ci.org/consul/consul.svg?branch=master)](https://travis-ci.org/consul/consul)
[![Code Climate](https://codeclimate.com/github/consul/consul/badges/gpa.svg)](https://codeclimate.com/github/consul/consul)
[![Coverage Status](https://coveralls.io/repos/github/consul/consul/badge.svg)](https://coveralls.io/github/consul/consul?branch=master)
[![Crowdin](https://d322cqt584bo4o.cloudfront.net/consul/localized.svg)](https://crowdin.com/project/consul)
[![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-blue.svg)](http://www.gnu.org/licenses/agpl-3.0)
[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)

[![Accessibility conformance](https://img.shields.io/badge/accessibility-WAI:AA-green.svg)](https://www.w3.org/WAI/eval/Overview)
[![A11y issues checked with Rocket Validator](https://rocketvalidator.com/badges/checked_with_rocket_validator.svg?url=https://rocketvalidator.com)](https://rocketvalidator.com/opensource)

[![Join the chat at https://gitter.im/consul/consul](https://badges.gitter.im/consul/consul.svg)](https://gitter.im/consul/consul?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://github.com/Ariwonto/consul/issues?q=is%3Aissue+is%3Aopen+label%3A"help+wanted")

This is the opensource code repository of the eParticipation website CONSUL, originally developed for the Madrid City government eParticipation website

## Documentation

Check the ongoing documentation at [https://docs.consulproject.org](https://docs.consulproject.org) to learn more about how to start your own CONSUL fork, install it, customize it and learn to use it from an administrator/maintainer perspective.

## CONSUL Project main website

You can access the main website of the project at [http://consulproject.org](http://consulproject.org) where you can find documentation about the use of the platform, videos, and links to the community space.

## Configuration for development and test environments

**NOTE**: For more detailed instructions check the [docs](https://docs.consulproject.org)

Prerequisites: install git, Ruby 2.4.9, `bundler` gem, Node.js and PostgreSQL (>=9.4).

```bash
git clone https://github.com/Ariwonto/consul.git
cd consul
bundle install
cp config/database.yml.example config/database.yml
cp config/secrets.yml.example config/secrets.yml
bin/rake db:create
bin/rake db:migrate
bin/rake db:dev_seed
RAILS_ENV=test rake db:setup
```

Run the app locally:

```bash
bin/rails s
```

Prerequisites for testing: install ChromeDriver >= 2.33

Run the tests with:

```bash
bin/rspec
```

You can use the default admin user from the seeds file:

 **user:** admin@politech.dev
 **pass:** 12345678

But for some actions like voting, you will need a verified user, the seeds file also includes one:

 **user:** verified@politech.dev
 **pass:** 12345678

## Configuration for production environments

See [installer](https://github.com/consul/installer)

## Current state

Development started on [2015 July 15th](https://github.com/Ariwonto/consul/commit/8db36308379accd44b5de4f680a54c41a0cc6fc6). Code was deployed to production on 2015 september 7th to [decide.madrid.es](https://decide.madrid.es). Since then new features are added often. You can take a look at the current features at the [project's website](http://consulproject.org/) and future features at the [Roadmap](https://github.com/Ariwonto/consul/projects/6) and [open issues list](https://github.com/Ariwonto/consul/issues).

## License

Code published under AFFERO GPL v3 (see [LICENSE-AGPLv3.txt](LICENSE-AGPLv3.txt))

## Contributions

See [CONTRIBUTING.md](CONTRIBUTING.md)

## add IOTA


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)

## [1.0.0](https://github.com/Ariwonto/consul/tree/1.0.0) (2019-06-10)
[Full Changelog](https://github.com/Ariwonto/consul/compare/1.0.0-beta...1.0.0)

### Added
- **Accounts:** Add description field to administrator users like evaluators description [\#3389](https://github.com/Ariwonto/consul/pull/3389)
- **Admin:** Add document uploads from admin section [\#3466](https://github.com/Ariwonto/consul/pull/3466)
- **Admin:** Images and documents settings [\#3585](https://github.com/Ariwonto/consul/pull/3585)
- **Budgets:** notify by email new evaluation comments [\#3413](https://github.com/Ariwonto/consul/pull/3413)
- **Installation:** Add deploy-secrets.yml.example file [\#3516](https://github.com/Ariwonto/consul/pull/3516)
- **Installation:** Add new settings automatically on every deployment [\#3576](https://github.com/Ariwonto/consul/pull/3576)
- **Installation:** Add task to upgrade to a new release [\#3590](https://github.com/Ariwonto/consul/pull/3590)
- **Legislations:** Create Legislation::PeopleProposal model [\#3591](https://github.com/Ariwonto/consul/pull/3591)
- **Translations:** Update translations from Crowdin [\#3378](https://github.com/Ariwonto/consul/pull/3378)
- **Translations:** Admin basic customization texts [\#3488](https://github.com/Ariwonto/consul/pull/3488)
- **Translations:** Add Bosnian, Croatian, Czech, Danish, Greek, and Turkish locales [\#3571](https://github.com/Ariwonto/consul/pull/3571)
- **Newsletters:** Proposals authors user segment [\#3507](https://github.com/Ariwonto/consul/pull/3507)
- **Polls:** Add slug to polls [\#3504](https://github.com/Ariwonto/consul/pull/3504)
- **Statistics:** Add budget stats [\#3438](https://github.com/Ariwonto/consul/pull/3438)
- **Statistics:** Add admin budget stats [\#3499](https://github.com/Ariwonto/consul/pull/3499)
- **Statistics:** Add options to show advanced stats [\#3520](https://github.com/Ariwonto/consul/pull/3520)

### Changed
- **Accounts:** Change devise configuration [\#3561](https://github.com/Ariwonto/consul/pull/3561)
- **Admin:** Show count of votes associated to verified signatures [\#2616](https://github.com/Ariwonto/consul/pull/2616)
- **Budgets:** Don't destroy budgets with an associated poll [\#3492](https://github.com/Ariwonto/consul/pull/3492)
- **Budgets:** Add task to regenerate ballot\_lines\_count cache [\#3563](https://github.com/Ariwonto/consul/pull/3563)
- **Dashboard:** Hide polls created by users from proposals dashboard on admin poll index [\#3572](https://github.com/Ariwonto/consul/pull/3572)
- **Dashboard:** Allow users to delete dashboard polls [\#3574](https://github.com/Ariwonto/consul/pull/3574)
- **Maintenance:** Add Rails 5.1 compatibility [\#3562](https://github.com/Ariwonto/consul/pull/3562)
- **Maintenance:** Update migrations and schema file [\#3598](https://github.com/Ariwonto/consul/pull/3598)
- **Maintenance-Refactoring:** Refactor admin/debates and admin/comments to hidden [\#3376](https://github.com/Ariwonto/consul/pull/3376)
- **Maintenance-Refactoring:** Simplify stats caching [\#3510](https://github.com/Ariwonto/consul/pull/3510)
- **Maintenance-Refactoring:** Refactor gender and age stats methods [\#3511](https://github.com/Ariwonto/consul/pull/3511)
- **Maintenance-Refactoring:** Simplify link to poll [\#3519](https://github.com/Ariwonto/consul/pull/3519)
- **Maintenance-Refactoring:** Extract partial with mobile sticky content [\#3577](https://github.com/Ariwonto/consul/pull/3577)
- **Maintenance-Refactoring:** Use find instead of find by [\#3580](https://github.com/Ariwonto/consul/pull/3580)
- **Maintenance-Rubocop:** Allow lines to be 110 characters long by Rubocop [\#3529](https://github.com/Ariwonto/consul/pull/3529)
- **Maintenance-Seeds:** Simplify settings seeds [\#3564](https://github.com/Ariwonto/consul/pull/3564)
- **Polls:** Display all polls for current booth [\#3361](https://github.com/Ariwonto/consul/pull/3361)
- **Polls:** Allow delete polls with associated questions and answers [\#3476](https://github.com/Ariwonto/consul/pull/3476)
- **Polls:** Remove redirect for poll officers [\#3506](https://github.com/Ariwonto/consul/pull/3506)
- **Polls:** Remove token on views [\#3539](https://github.com/Ariwonto/consul/pull/3539)
- **Proposals:** Remove question and external_url fields from proposals and legislation proposals [\#3397](https://github.com/Ariwonto/consul/pull/3397)
- **Proposals:** Proposals support on mobile [\#3515](https://github.com/Ariwonto/consul/pull/3515)
- **Proposals:** Make proposals to be selected by administrators [\#3567](https://github.com/Ariwonto/consul/pull/3567)
- **Statistics:** Improve poll stats [\#3503](https://github.com/Ariwonto/consul/pull/3503)
- **Statistics:** Change stats layout [\#3512](https://github.com/Ariwonto/consul/pull/3512)
- **UX/UI:** Improve help texts on Admin UI [\#3508](https://github.com/Ariwonto/consul/pull/3508)
- **UX/UI:** Users menu [\#3509](https://github.com/Ariwonto/consul/pull/3509)
- **UX/UI:** Add help texs, links and new message section to improve UX [\#3573](https://github.com/Ariwonto/consul/pull/3573)

### Fixed
- **Budgets:** Don't show links to disabled budget results [\#3592](https://github.com/Ariwonto/consul/pull/3592)
- **Legislations:** Fix order in annotation comments with same score [\#3565](https://github.com/Ariwonto/consul/pull/3565)
- **Maintenance:** Fix obsolete `respond\_with\_bip` usage [\#3483](https://github.com/Ariwonto/consul/pull/3483)
- **Maintenance:** Remove Rspec deprecation warning [\#3530](https://github.com/Ariwonto/consul/pull/3530)
- **Maintenance:** Fix column order in schema file [\#3533](https://github.com/Ariwonto/consul/pull/3533)
- **Maintenance:** Fix indentation in schema file [\#3595](https://github.com/Ariwonto/consul/pull/3595)

spec/factories/users.rb
FactoryBot.define do
  factory :user do
    sequence(:username) { |n| "Manuela#{n}" }
    sequence(:email)    { |n| "manuela#{n}@politech.dev" }

    password            { "judgmentday" }
    terms_of_service    { "1" }
    confirmed_at        { Time.current }
    date_of_birth       { 20.years.ago }
    public_activity     { true }

    trait :incomplete_verification do
      after :create do |user|
        create(:failed_census_call, user: user)
      end
    end

    trait :level_two do
      residence_verified_at { Time.current }
      unconfirmed_phone { "611111111" }
      confirmed_phone { "611111111" }
      sms_confirmation_code { "1234" }
      document_type { "1" }
      document_number
      date_of_birth { Date.new(1980, 12, 31) }
      gender { "female" }
      geozone
    end

    trait :level_three do
      verified_at { Time.current }
      document_type { "1" }
      document_number
    end

    trait :hidden do
      hidden_at { Time.current }
    end

    trait :with_confirmed_hide do
      confirmed_hide_at { Time.current }
    end

    trait :verified do
      residence_verified_at { Time.current }
      verified_at { Time.current }
    end

    trait :in_census do
      document_number { "12345678Z" }
      document_type { "1" }
      verified_at { Time.current }
    end

    trait :with_proposal do
      after(:create) { |user| create(:proposal, author: user) }
    end

    trait :with_debate do
      after(:create) { |user| create(:debate, author: user) }
    end

    trait :with_comment do
      after(:create) { |user| create(:comment, author: user) }
    end

    transient do
      votables { [] }
      followables { [] }
      ballot_lines { [] }
    end

    after(:create) do |user, evaluator|
      evaluator.votables.each { |votable| create(:vote, votable: votable, voter: user) }
      evaluator.followables.each { |followable| create(:follow, followable: followable, user: user) }

      if evaluator.ballot_lines.any?
        ballot = create(:budget_ballot, budget: evaluator.ballot_lines.first.budget.reload, user: user)

        evaluator.ballot_lines.each do |investment|
          create(:budget_ballot_line, investment: investment, ballot: ballot)
        end
      end
    end
  end

  factory :identity do
    user { nil }
    provider { "Twitter" }
    uid { "MyString" }
  end

  factory :administrator do
    user
  end

  factory :moderator do
    user
  end

  factory :valuator do
    user
  end

  factory :manager do
    user
  end

  factory :poll_officer, class: "Poll::Officer" do
    user { association(:user, username: name) }

    transient do
      sequence(:name) { |n| "Officer #{n}" }
      polls { [] }
    end

    after(:create) do |officer, evaluator|
      evaluator.polls.each do |poll|
        create(:poll_officer_assignment, poll: poll, officer: officer)
      end
    end
  end

  factory :follow do
    association :user, factory: :user

    trait :followed_proposal do
      association :followable, factory: :proposal
    end

    trait :followed_investment do
      association :followable, factory: :budget_investment
    end
  end

  factory :direct_message do
    title    { "Hey" }
    body     { "How are You doing?" }
    association :sender,   factory: :user
    association :receiver, factory: :user
  end
end


spec/factories/emails.rb
FactoryBot.define do
  factory :newsletter do
    sequence(:subject) { |n| "Subject #{n}" }
    segment_recipient  { UserSegments::SEGMENTS.sample }
    sequence(:from)    { |n| "noreply#{n}@politech.dev" }
    sequence(:body)    { |n| "Body #{n}" }
  end
end


config/deploy.rb
# config valid only for current version of Capistrano
lock "~> 3.10.1"

def deploysecret(key)
  @deploy_secrets_yml ||= YAML.load_file("config/deploy-secrets.yml")[fetch(:stage).to_s]
  @deploy_secrets_yml.fetch(key.to_s, "undefined")
end

set :rails_env, fetch(:stage)
set :rvm1_map_bins, -> { fetch(:rvm_map_bins).to_a.concat(%w[rake gem bundle ruby]).uniq }

set :application, "consul"
set :full_app_name, deploysecret(:full_app_name)
set :deploy_to, deploysecret(:deploy_to)
set :server_name, deploysecret(:server_name)
set :db_server, deploysecret(:db_server)
set :ssh_options, port: deploysecret(:ssh_port)

set :repo_url, "https://github.com/Ariwonto/consul.git"

set :revision, `git rev-parse --short #{fetch(:branch)}`.strip

set :log_level, :info
set :pty, true
set :use_sudo, false

set :linked_files, %w[config/database.yml config/secrets.yml]
set :linked_dirs, %w[log tmp public/system public/assets public/ckeditor_assets]

set :keep_releases, 5

set :local_user, ENV["USER"]

set :puma_conf, "#{release_path}/config/puma/#{fetch(:rails_env)}.rb"

set :delayed_job_workers, 2
set :delayed_job_roles, :background

set(:config_files, %w[
  log_rotation
  database.yml
  secrets.yml
])

set :whenever_roles, -> { :app }

namespace :deploy do
  after :updating, "rvm1:install:rvm"
  after :updating, "rvm1:install:ruby"
  after :updating, "install_bundler_gem"
  before "deploy:migrate", "remove_local_census_records_duplicates"

  after "deploy:migrate", "add_new_settings"

  before :publishing, "smtp_ssl_and_delay_jobs_secrets"
  after  :publishing, "setup_puma"

  after :published, "deploy:restart"
  before "deploy:restart", "puma:smart_restart"
  before "deploy:restart", "delayed_job:restart"

  after :finished, "refresh_sitemap"

  desc "Deploys and runs the tasks needed to upgrade to a new release"
  task :upgrade do
    after "add_new_settings", "execute_release_tasks"
    invoke "deploy"
  end
end

task :install_bundler_gem do
  on roles(:app) do
    within release_path do
      execute :rvm, fetch(:rvm1_ruby_version), "do", "gem install bundler"
    end
  end
end

task :remove_local_census_records_duplicates do
  on roles(:db) do
    within release_path do
      with rails_env: fetch(:rails_env) do
        execute :rake, "local_census_records:remove_duplicates"
      end
    end
  end
end

task :refresh_sitemap do
  on roles(:app) do
    within release_path do
      with rails_env: fetch(:rails_env) do
        execute :rake, "sitemap:refresh:no_ping"
      end
    end
  end
end

task :add_new_settings do
  on roles(:db) do
    within release_path do
      with rails_env: fetch(:rails_env) do
        execute :rake, "settings:add_new_settings"
      end
    end
  end
end

task :execute_release_tasks do
  on roles(:app) do
    within release_path do
      with rails_env: fetch(:rails_env) do
        execute :rake, "consul:execute_release_tasks"
      end
    end
  end
end

desc "Create pid and socket folders needed by puma and convert unicorn sockets into symbolic links \
      to the puma socket, so legacy nginx configurations pointing to the unicorn socket keep working"
task :setup_puma do
  on roles(:app) do
    with rails_env: fetch(:rails_env) do
      execute "mkdir -p #{shared_path}/tmp/sockets; true"
      execute "mkdir -p #{shared_path}/tmp/pids; true"

      if test("[ -e #{shared_path}/tmp/sockets/unicorn.sock ]")
        execute "ln -sf #{shared_path}/tmp/sockets/puma.sock #{shared_path}/tmp/sockets/unicorn.sock; true"
      end

      if test("[ -e #{shared_path}/sockets/unicorn.sock ]")
        execute "ln -sf #{shared_path}/tmp/sockets/puma.sock #{shared_path}/sockets/unicorn.sock; true"
      end
    end
  end
end

task :smtp_ssl_and_delay_jobs_secrets do
  on roles(:app) do
    if test("[ -d #{current_path} ]")
      within current_path do
        with rails_env: fetch(:rails_env) do
          tasks_file_path = "lib/tasks/secrets.rake"
          shared_secrets_path = "#{shared_path}/config/secrets.yml"

          unless test("[ -e #{current_path}/#{tasks_file_path} ]")
            begin
              unless test("[ -w #{shared_secrets_path} ]")
                execute "sudo chown `whoami` #{shared_secrets_path}"
                execute "chmod u+w #{shared_secrets_path}"
              end

              execute "cp #{release_path}/#{tasks_file_path} #{current_path}/#{tasks_file_path}"

              execute :rake, "secrets:smtp_ssl_and_delay_jobs"
            ensure
              execute "rm #{current_path}/#{tasks_file_path}"
            end
          end
        end
      end
    end
  end
end


app/models/setting.rb
class Setting < ApplicationRecord
  validates :key, presence: true, uniqueness: true

  default_scope { order(id: :asc) }

  def prefix
    key.split(".").first
  end

  def type
    if %w[feature process proposals map html homepage uploads].include? prefix
      prefix
    elsif %w[remote_census].include? prefix
      key.rpartition(".").first
    else
      "configuration"
    end
  end

  def enabled?
    value.present?
  end

  def content_type?
    key.split(".").last == "content_types"
  end

  def content_type_group
    key.split(".").second
  end

  class << self
    def [](key)
      where(key: key).pluck(:value).first.presence
    end

    def []=(key, value)
      setting = find_by(key: key) || new(key: key)
      setting.value = value.presence
      setting.save!
      value
    end

    def rename_key(from:, to:)
      if where(key: to).empty?
        value = where(key: from).pluck(:value).first.presence
        create!(key: to, value: value)
      end
      remove(from)
    end

    def remove(key)
      setting = find_by(key: key)
      setting.destroy if setting.present?
    end

    def accepted_content_types_for(group)
      mime_content_types = Setting["uploads.#{group}.content_types"]&.split(" ") || []
      Setting.mime_types[group].select { |_, content_type| mime_content_types.include?(content_type) }.keys
    end

    def mime_types
      {
        "images" => {
          "jpg"  => "image/jpeg",
          "png"  => "image/png",
          "gif"  => "image/gif"
        },
        "documents" => {
          "pdf"  => "application/pdf",
          "doc"  => "application/msword",
          "docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
          "xls"  => "application/x-ole-storage",
          "xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
          "csv"  => "text/plain",
          "zip"  => "application/zip"
        }
      }
    end

    def defaults
      {
        "feature.featured_proposals": nil,
        "feature.facebook_login": true,
        "feature.google_login": true,
        "feature.twitter_login": true,
        "feature.public_stats": true,
        "feature.signature_sheets": true,
        "feature.user.recommendations": true,
        "feature.user.recommendations_on_debates": true,
        "feature.user.recommendations_on_proposals": true,
        "feature.user.skip_verification": "true",
        "feature.community": true,
        "feature.map": nil,
        "feature.allow_attached_documents": true,
        "feature.allow_images": true,
        "feature.help_page": true,
        "feature.remote_translations": nil,
        "feature.translation_interface": nil,
        "feature.remote_census": nil,
        "feature.valuation_comment_notification": true,
        "feature.graphql_api": true,
        "homepage.widgets.feeds.debates": true,
        "homepage.widgets.feeds.processes": true,
        "homepage.widgets.feeds.proposals": true,
        # Code to be included at the top (inside <body>) of every page
        "html.per_page_code_body": "",
        # Code to be included at the top (inside <head>) of every page (useful for tracking)
        "html.per_page_code_head": "",
        "map.latitude": 51.48,
        "map.longitude": 0.0,
        "map.zoom": 10,
        "process.debates": true,
        "process.proposals": true,
        "process.polls": true,
        "process.budgets": true,
        "process.legislation": true,
        "proposals.successful_proposal_id": nil,
        "proposals.poll_short_title": nil,
        "proposals.poll_description": nil,
        "proposals.poll_link": nil,
        "proposals.email_short_title": nil,
        "proposals.email_description": nil,
        "proposals.poster_short_title": nil,
        "proposals.poster_description": nil,
        # Images and Documents
        "uploads.images.title.min_length": 4,
        "uploads.images.title.max_length": 80,
        "uploads.images.min_width": 0,
        "uploads.images.min_height": 475,
        "uploads.images.max_size": 1,
        "uploads.images.content_types": "image/jpeg",
        "uploads.documents.max_amount": 3,
        "uploads.documents.max_size": 3,
        "uploads.documents.content_types": "application/pdf",
        # Names for the moderation console, as a hint for moderators
        # to know better how to assign users with official positions
        "official_level_1_name": I18n.t("seeds.settings.official_level_1_name"),
        "official_level_2_name": I18n.t("seeds.settings.official_level_2_name"),
        "official_level_3_name": I18n.t("seeds.settings.official_level_3_name"),
        "official_level_4_name": I18n.t("seeds.settings.official_level_4_name"),
        "official_level_5_name": I18n.t("seeds.settings.official_level_5_name"),
        "max_ratio_anon_votes_on_debates": 50,
        "max_votes_for_debate_edit": 1000,
        "max_votes_for_proposal_edit": 1000,
        "comments_body_max_length": 1000,
        "proposal_code_prefix": "CONSUL",
        "votes_for_proposal_success": 10000,
        "months_to_archive_proposals": 12,
        # Users with this email domain will automatically be marked as level 1 officials
        # Emails under the domain's subdomains will also be included
        "email_domain_for_officials": "",
        "facebook_handle": nil,
        "instagram_handle": nil,
        "telegram_handle": nil,
        "twitter_handle": nil,
        "twitter_hashtag": nil,
        "youtube_handle": nil,
        "url": "http://example.com", # Public-facing URL of the app.
        # CONSUL installation's organization name
        "org_name": "CONSUL",
        "meta_title": nil,
        "meta_description": nil,
        "meta_keywords": nil,
        "proposal_notification_minimum_interval_in_days": 3,
        "direct_message_max_per_day": 3,
        "mailer_from_name": "CONSUL",
        "mailer_from_address": "noreply@politech.dev",
        "min_age_to_participate": 16,
        "hot_score_period_in_days": 31,
        "related_content_score_threshold": -0.3,
        "featured_proposals_number": 3,
        "dashboard.emails": nil,
        "remote_census.general.endpoint": "",
        "remote_census.request.method_name": "",
        "remote_census.request.structure": "",
        "remote_census.request.document_type": "",
        "remote_census.request.document_number": "",
        "remote_census.request.date_of_birth": "",
        "remote_census.request.postal_code": "",
        "remote_census.response.date_of_birth": "",
        "remote_census.response.postal_code": "",
        "remote_census.response.district": "",
        "remote_census.response.gender": "",
        "remote_census.response.name": "",
        "remote_census.response.surname": "",
        "remote_census.response.valid": ""
      }
    end

    def reset_defaults
      defaults.each { |name, value| self[name] = value }
    end

    def add_new_settings
      defaults.each do |name, value|
        self[name] = value unless find_by(key: name)
      end
    end

    def force_presence_date_of_birth?
      Setting["feature.remote_census"].present? &&
        Setting["remote_census.request.date_of_birth"].present?
    end

    def force_presence_postal_code?
      Setting["feature.remote_census"].present? &&
        Setting["remote_census.request.postal_code"].present?
    end
  end
end


config/application.rb
require_relative "boot"

require "rails/all"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module Consul
  class Application < Rails::Application
    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

    # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
    # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
    # config.time_zone = 'Central Time (US & Canada)'

    # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
    # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}').to_s]
    config.i18n.default_locale = :en
    available_locales = [
      "ar",
      "bs",
      "cs",
      "da",
      "de",
      "el",
      "en",
      "es",
      "fa",
      "fr",
      "gl",
      "he",
      "hr",
      "id",
      "it",
      "nl",
      "pl",
      "pt-BR",
      "ru",
      "sl",
      "sq",
      "so",
      "sv",
      "tr",
      "val",
      "zh-CN",
      "zh-TW"]
    config.i18n.available_locales = available_locales
    config.i18n.fallbacks = {
      "fr"    => "es",
      "gl"    => "es",
      "it"    => "es",
      "pt-BR" => "es"
    }

    config.i18n.load_path += Dir[Rails.root.join("config", "locales", "**[^custom]*", "*.{rb,yml}")]
    config.i18n.load_path += Dir[Rails.root.join("config", "locales", "custom", "**", "*.{rb,yml}")]

    config.after_initialize do
      Globalize.set_fallbacks_to_all_available_locales
      GraphQLApi::Loader.setup
    end

    config.assets.paths << Rails.root.join("app", "assets", "fonts")
    config.assets.paths << Rails.root.join("vendor", "assets", "fonts")

    # Add lib to the autoload path
    config.autoload_paths << Rails.root.join("lib")
    config.time_zone = "Madrid"
    config.active_job.queue_adapter = :delayed_job

    # CONSUL specific custom overrides
    # Read more on documentation:
    # * English: https://github.com/Ariwonto/consul/blob/master/CUSTOMIZE_EN.md
    # * Spanish: https://github.com/Ariwonto/consul/blob/master/CUSTOMIZE_ES.md
    #
    config.autoload_paths << "#{Rails.root}/app/controllers/custom"
    config.autoload_paths << "#{Rails.root}/app/models/custom"
    config.paths["app/views"].unshift(Rails.root.join("app", "views", "custom"))
  end
end

class Rails::Engine
  initializer :prepend_custom_assets_path, group: :all do |app|
    if self.class.name == "Consul::Application"
      %w[images fonts javascripts].each do |asset|
        app.config.assets.paths.unshift(Rails.root.join("app", "assets", asset, "custom").to_s)
      end
    end
  end
end

require "./config/application_custom.rb"


db/dev_seeds/settings.rb
section "Creating Settings" do
  Setting.reset_defaults

  {
    "facebook_handle": "CONSUL",
    "feature.featured_proposals": "true",
    "feature.map": "true",
    "instagram_handle": "CONSUL",
    "mailer_from_address": "noreply@politech.dev",
    "mailer_from_name": "CONSUL",
    "meta_description": "Citizen participation tool for an open, "\
                        "transparent and democratic government",
    "meta_keywords": "citizen participation, open government",
    "meta_title": "CONSUL",
    "org_name": "CONSUL",
    "proposal_code_prefix": "MAD",
    "proposal_notification_minimum_interval_in_days": 0,
    "telegram_handle": "CONSUL",
    "twitter_handle": "@consul_dev",
    "twitter_hashtag": "#consul_dev",
    "url": "http://localhost:3000",
    "votes_for_proposal_success": "100",
    "youtube_handle": "CONSUL"
  }.each do |name, value|
    Setting[name] = value
  end
end


db/dev_seeds/newsletters.rb
section "Creating Newsletters" do
  newsletter_body = [
    "We choose to go to the moon in this decade and do the other things, not because they are easy"\
    ", but because they are hard, because that goal will serve to organize and measure the best of"\
    " our energies and skills, because that challenge is one that we are willing to accept, one we"\
    " are unwilling to postpone, and one which we intend to win.",
    "Spaceflights cannot be stopped. This is not the work of any one man or even a group of men."\
    " It is a historical process which mankind is carrying out in accordance with the natural laws"\
    " of human development.",
    "Many say exploration is part of our destiny, but it’s actually our duty to future generations"\
    " and their quest to ensure the survival of the human species."
  ]

  5.times do |n|
    Newsletter.create!(
      subject: "Newsletter subject #{n}",
      segment_recipient: UserSegments::SEGMENTS.sample,
      from: "no-reply@politech.dev",
      body: newsletter_body.sample,
      sent_at: [Time.current, nil].sample
    )
  end
end


db/dev_seeds/users.rb
section "Creating Users" do
  def create_user(email, username)
    password = "12345678"
    User.create!(
      username:               username,
      email:                  email,
      password:               password,
      password_confirmation:  password,
      confirmed_at:           Time.current,
      terms_of_service:       "1",
      gender:                 %w[male female].sample,
      date_of_birth:          rand((Time.current - 80.years)..(Time.current - 16.years)),
      public_activity:        (rand(1..100) > 30)
    )
  end

  def unique_document_number
    @document_number ||= 12345678
    @document_number += 1
    "#{@document_number}#{[*"A".."Z"].sample}"
  end

  admin = create_user("admin@politech.dev", "admin")
  admin.create_administrator
  admin.update!(residence_verified_at: Time.current,
               confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
               verified_at: Time.current, document_number: unique_document_number)

  moderator = create_user("mod@politech.dev", "moderator")
  moderator.create_moderator
  moderator.update!(residence_verified_at: Time.current,
                   confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                   verified_at: Time.current, document_number: unique_document_number)

  manager = create_user("manager@politech.dev", "manager")
  manager.create_manager
  manager.update!(residence_verified_at: Time.current,
                 confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                 verified_at: Time.current, document_number: unique_document_number)

  valuator = create_user("valuator@politech.dev", "valuator")
  valuator.create_valuator
  valuator.update!(residence_verified_at: Time.current,
                  confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                  verified_at: Time.current, document_number: unique_document_number)

  poll_officer = create_user("poll_officer@politech.dev", "Paul O. Fisher")
  poll_officer.create_poll_officer
  poll_officer.update!(residence_verified_at: Time.current,
                      confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                      verified_at: Time.current, document_number: unique_document_number)

  poll_officer2 = create_user("poll_officer2@politech.dev", "Pauline M. Espinosa")
  poll_officer2.create_poll_officer
  poll_officer2.update!(residence_verified_at: Time.current,
                       confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                       verified_at: Time.current, document_number: unique_document_number)

  create_user("unverified@politech.dev", "unverified")

  level_2 = create_user("leveltwo@politech.dev", "level 2")
  level_2.update!(residence_verified_at: Time.current,
                 confirmed_phone: Faker::PhoneNumber.phone_number,
                 document_number: unique_document_number, document_type: "1")

  verified = create_user("verified@politech.dev", "verified")
  verified.update!(residence_verified_at: Time.current,
                  confirmed_phone: Faker::PhoneNumber.phone_number, document_type: "1",
                  verified_at: Time.current, document_number: unique_document_number)

  [
    I18n.t("seeds.organizations.neighborhood_association"),
    I18n.t("seeds.organizations.human_rights"),
    "Greenpeace"
  ].each do |organization_name|
    org_user = create_user("#{organization_name.parameterize}@politech.dev", organization_name)
    org = org_user.create_organization(name: organization_name, responsible_name: Faker::Name.name)
    [true, false].cycle ? org.verify : org.reject
  end

  5.times do |i|
    official = create_user("official#{i}@politech.dev", "Official #{i}")
    official.update!(official_level: i, official_position: "Official position #{i}")
  end

  30.times do |i|
    user = create_user("user#{i}@politech.dev", "Regular user #{i}")
    level = [1, 2, 3].sample
    if level >= 2
      user.update(residence_verified_at: Time.current,
                  confirmed_phone: Faker::PhoneNumber.phone_number,
                  document_number: unique_document_number,
                  document_type: "1",
                  geozone: Geozone.all.sample)
    end
    if level == 3
      user.update(verified_at: Time.current, document_number: unique_document_number)
    end
  end
end
